File: blueprints/search/pkg/engine/local/engines/arxiv.go
package engines

import (
	"context"
	"encoding/xml"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// ArXiv implements arXiv academic paper search.
type ArXiv struct {
	*BaseEngine
}

// NewArXiv creates a new ArXiv engine.
func NewArXiv() *ArXiv {
	a := &ArXiv{
		BaseEngine: NewBaseEngine("arxiv", "arx", []Category{CategoryScience}),
	}

	a.SetPaging(true).
		SetMaxPage(10).
		SetTimeout(5 * time.Second).
		SetAbout(EngineAbout{
			Website:         "https://arxiv.org",
			WikidataID:      "Q118398",
			OfficialAPIDocs: "https://info.arxiv.org/help/api/index.html",
			UseOfficialAPI:  true,
			Results:         "XML",
		})

	return a
}

func (a *ArXiv) Request(ctx context.Context, query string, params *RequestParams) error {
	maxResults := 10
	start := 0
	if params.PageNo > 1 {
		start = (params.PageNo - 1) * maxResults
	}

	queryParams := url.Values{}
	queryParams.Set("search_query", "all:"+query)
	queryParams.Set("start", fmt.Sprintf("%d", start))
	queryParams.Set("max_results", fmt.Sprintf("%d", maxResults))

	params.URL = "https://export.arxiv.org/api/query?" + queryParams.Encode()
	params.Headers.Set("Accept", "application/atom+xml")

	return nil
}

func (a *ArXiv) Response(ctx context.Context, resp *http.Response, params *RequestParams) (*EngineResults, error) {
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	results := NewEngineResults()

	var feed arxivFeed
	if err := xml.Unmarshal(body, &feed); err != nil {
		return nil, err
	}

	for _, entry := range feed.Entries {
		result := Result{
			URL:      entry.ID,
			Title:    strings.TrimSpace(entry.Title),
			Content:  strings.TrimSpace(entry.Summary),
			Template: "paper",
		}

		// Extract authors
		authors := make([]string, 0, len(entry.Authors))
		for _, author := range entry.Authors {
			authors = append(authors, author.Name)
		}
		result.Authors = authors

		// Parse published date
		if entry.Published != "" {
			if t, err := time.Parse(time.RFC3339, entry.Published); err == nil {
				result.PublishedAt = t
			}
		}

		// Find PDF link
		for _, link := range entry.Links {
			if link.Title == "pdf" {
				// Store PDF URL (could use a custom field or content)
				result.Content = result.Content + " [PDF: " + link.Href + "]"
			}
		}

		// Extract DOI if present
		if entry.DOI != "" {
			result.DOI = entry.DOI
		}

		// Extract journal reference
		if entry.JournalRef != "" {
			result.Journal = entry.JournalRef
		}

		result.ParsedURL, _ = url.Parse(result.URL)
		results.Add(result)
	}

	return results, nil
}

type arxivFeed struct {
	XMLName xml.Name     `xml:"feed"`
	Entries []arxivEntry `xml:"entry"`
}

type arxivEntry struct {
	ID         string       `xml:"id"`
	Title      string       `xml:"title"`
	Summary    string       `xml:"summary"`
	Published  string       `xml:"published"`
	Updated    string       `xml:"updated"`
	Authors    []arxivAuthor `xml:"author"`
	Links      []arxivLink  `xml:"link"`
	DOI        string       `xml:"doi"`
	JournalRef string       `xml:"journal_ref"`
	Comment    string       `xml:"comment"`
	Categories []struct {
		Term string `xml:"term,attr"`
	} `xml:"category"`
}

type arxivAuthor struct {
	Name string `xml:"name"`
}

type arxivLink struct {
	Href  string `xml:"href,attr"`
	Title string `xml:"title,attr"`
	Type  string `xml:"type,attr"`
}

File: blueprints/search/app/worker/src/engines/arxiv.ts
/**
 * arXiv Search Engine adapter.
 * Ported from Go: pkg/engine/local/engines/arxiv.go
 *
 * Uses export.arxiv.org/api/query XML (ATOM) endpoint.
 */

import type {
  OnlineEngine,
  EngineParams,
  RequestConfig,
  EngineResults,
  Category,
} from './engine';
import { newEngineResults } from './engine';
import {
  getElementsByTagName,
  getTextContent,
  getElementAttribute,
} from '../lib/xml-parser';

export class ArxivEngine implements OnlineEngine {
  name = 'arxiv';
  shortcut = 'arx';
  categories: Category[] = ['science'];
  supportsPaging = true;
  maxPage = 10;
  timeout = 5_000;
  weight = 1.0;
  disabled = false;

  buildRequest(query: string, params: EngineParams): RequestConfig {
    const maxResults = 10;
    let start = 0;
    if (params.page > 1) {
      start = (params.page - 1) * maxResults;
    }

    const searchParams = new URLSearchParams();
    searchParams.set('search_query', `all:${query}`);
    searchParams.set('start', start.toString());
    searchParams.set('max_results', maxResults.toString());

    return {
      url: `https://export.arxiv.org/api/query?${searchParams.toString()}`,
      method: 'GET',
      headers: {
        Accept: 'application/atom+xml',
      },
      cookies: [],
    };
  }

  parseResponse(body: string, _params: EngineParams): EngineResults {
    const results = newEngineResults();

    // Parse ATOM XML entries
    const entries = getElementsByTagName(body, 'entry');

    for (const entry of entries) {
      // Extract ID (URL)
      const id = getTextContent(entry, 'id').trim();
      if (!id) continue;

      // Extract title
      const title = getTextContent(entry, 'title')
        .replace(/\s+/g, ' ')
        .trim();

      // Extract summary
      const summary = getTextContent(entry, 'summary')
        .replace(/\s+/g, ' ')
        .trim();

      // Extract authors
      const authorNames = getElementsByTagName(entry, 'author');
      const authors: string[] = [];
      for (const authorXml of authorNames) {
        const name = getTextContent(authorXml, 'name').trim();
        if (name) {
          authors.push(name);
        }
      }

      // Extract published date
      const published = getTextContent(entry, 'published').trim();
      let publishedAt = '';
      if (published) {
        try {
          publishedAt = new Date(published).toISOString();
        } catch {
          publishedAt = published;
        }
      }

      // Find PDF link
      const linkHrefs = getElementAttribute(entry, 'link', 'href');
      const linkTitles = getElementAttribute(entry, 'link', 'title');
      let pdfUrl = '';
      for (let i = 0; i < linkTitles.length; i++) {
        if (linkTitles[i] === 'pdf' && linkHrefs[i]) {
          pdfUrl = linkHrefs[i];
          break;
        }
      }

      // Extract DOI
      const doi = getTextContent(entry, 'doi').trim();

      // Extract journal reference
      const journal = getTextContent(entry, 'journal_ref').trim();

      // Build content with PDF link appended
      let content = summary;
      if (pdfUrl) {
        content += ` [PDF: ${pdfUrl}]`;
      }

      results.results.push({
        url: id,
        title,
        content,
        engine: this.name,
        score: this.weight,
        category: 'science',
        template: 'paper',
        authors,
        publishedAt,
        doi: doi || undefined,
        journal: journal || undefined,
      });
    }

    return results;
  }
}
